### ISSUE TITLE
Launching Jupyter notebook service fails

### ISSUE THREAD

**Issue by gabrielgrant** [pachyderm/pachyderm#2846]
Using a spec basically directly from http://docs.pachyderm.io/en/latest/fundamentals/creating_services.html :

```
{
    "pipeline" : {
      "name": "jupyter"
    },
    "input": {
        "atom": {
            "glob": "/",
            "repo": "testrepo",
            "branch": "master"
        }
    },
    "service": {
        "external_port": 30888,
        "internal_port": 8888
    },
    "transform": {
        "cmd": [
            "start-notebook.sh"
        ],
        "image": "jupyter/datascience-notebook"
    }
}
```

creating seems to succeed, but it gets stuck in "starting" state:

```
$ pc list-pipeline
NAME                             INPUT                              OUTPUT                           CREATED                    STATE
jupyter                          testrepo:/                         jupyter/master                   28 minutes ago             starting  
```

Problem seems to be: 
```
$ kubectl logs po/pipeline-jupyter-v1-n8kzm user
user 1000 not found
```

Pod details:

```
$ kubectl describe po/pipeline-jupyter-v1-n8kzm
Name:       pipeline-jupyter-v1-n8kzm
Namespace:  default
Node:       rexy2/192.168.1.128
Start Time: Fri, 13 Apr 2018 14:23:00 -0700
Labels:     app=pipeline-jupyter-v1
        component=worker
        pipelineName=jupyter
        suite=pachyderm
        version=1.7.0-64b16d2a615d5e6cbb3b82a9d3c689f076c3e44f
Annotations:    kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicationController","namespace":"default","name":"pipeline-jupyter-v1","uid":"d7b038c5-3f60-11e8-acd9-f...
        pipelineName=jupyter
Status:     Running
IP:     172.17.0.8
Controllers:    ReplicationController/pipeline-jupyter-v1
Init Containers:
  init:
    Container ID:   docker://243fa4f43adab37ce1dcff17a5fc9310f70edd44442d2b025b567fa31ea57521
    Image:      pachyderm/worker:local
    Image ID:       docker://sha256:6eebe847dc8618e2d31d6cd1e5d3adfe349a65fac6af10317d4331d9c73da3e7
    Port:       
    Command:
      /pach/worker.sh
    State:      Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Fri, 13 Apr 2018 14:23:01 -0700
      Finished:     Fri, 13 Apr 2018 14:23:01 -0700
    Ready:      True
    Restart Count:  0
    Environment:
      PPS_WORKER_IP:         (v1:status.podIP)
      PPS_POD_NAME:     pipeline-jupyter-v1-n8kzm (v1:metadata.name)
      PPS_ETCD_PREFIX:      pachyderm/1.7.0/pachyderm_pps
      PPS_NAMESPACE:        default
      PPS_SPEC_COMMIT:      a358d2e55c144dc292797137cc3b7df3
      PPS_PIPELINE_NAME:    jupyter
    Mounts:
      /pach-bin from pach-bin (rw)
      /pachyderm-storage-secret from pachyderm-storage-secret (rw)
      /scratch from pachyderm-worker (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from pachyderm-token-82jpz (ro)
Containers:
  user:
    Container ID:   docker://347366959b7722415693d9863ac99218c1ac9ac8f05b04298fafb746160f1571
    Image:      jupyter/datascience-notebook
    Image ID:       docker-pullable://jupyter/datascience-notebook@sha256:22a5429fc41df4bf2a666000190685c92fbd538e111d3d944320c0f868c45cbb
    Port:       
    Command:
      /pach-bin/guest.sh
    State:      Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Fri, 13 Apr 2018 14:47:42 -0700
      Finished:     Fri, 13 Apr 2018 14:47:42 -0700
    Ready:      False
    Restart Count:  7
    Requests:
      cpu:  0
      memory:   64M
    Environment:
      PPS_WORKER_IP:         (v1:status.podIP)
      PPS_POD_NAME:     pipeline-jupyter-v1-n8kzm (v1:metadata.name)
      PPS_ETCD_PREFIX:      pachyderm/1.7.0/pachyderm_pps
      PPS_NAMESPACE:        default
      PPS_SPEC_COMMIT:      a358d2e55c144dc292797137cc3b7df3
      PPS_PIPELINE_NAME:    jupyter
    Mounts:
      /pach-bin from pach-bin (rw)
      /pachyderm-storage-secret from pachyderm-storage-secret (rw)
      /scratch from pachyderm-worker (rw)
      /var/run/docker.sock from docker (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from pachyderm-token-82jpz (ro)
  storage:
    Container ID:   docker://a3d5917a2decca765e3a98699092ff336b0d0042f946b85561c850345e666938
    Image:      pachyderm/pachd:local
    Image ID:       docker://sha256:fd01c8b8c9cd606287827c2994b91c395184055315bc4265a27e3dc7050ce462
    Port:       
    Command:
      /pachd
      --mode
      sidecar
    State:      Running
      Started:      Fri, 13 Apr 2018 14:36:40 -0700
    Ready:      True
    Restart Count:  0
    Requests:
      cpu:  0
      memory:   64M
    Environment:
      BLOCK_CACHE_BYTES:    64M
      PFS_CACHE_SIZE:       16
      PACH_ROOT:        /pach
      STORAGE_BACKEND:      LOCAL
    Mounts:
      /pach from pach-disk (rw)
      /pachyderm-storage-secret from pachyderm-storage-secret (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from pachyderm-token-82jpz (ro)
Conditions:
  Type      Status
  Initialized   True 
  Ready     False 
  PodScheduled  True 
Volumes:
  pach-bin:
    Type:   EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium: 
  pachyderm-worker:
    Type:   EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium: 
  pach-disk:
    Type:   HostPath (bare host directory volume)
    Path:   /var/pachyderm/pachd
  pachyderm-storage-secret:
    Type:   Secret (a volume populated by a Secret)
    SecretName: pachyderm-storage-secret
    Optional:   false
  docker:
    Type:   HostPath (bare host directory volume)
    Path:   /var/run/docker.sock
  pachyderm-token-82jpz:
    Type:   Secret (a volume populated by a Secret)
    SecretName: pachyderm-token-82jpz
    Optional:   false
QoS Class:  Burstable
Node-Selectors: <none>
Tolerations:    <none>
Events:
  FirstSeen LastSeen    Count   From            SubObjectPath           Type        Reason          Message
  --------- --------    -----   ----            -------------           --------    ------          -------
  28m       28m     1   default-scheduler                   Normal      Scheduled       Successfully assigned pipeline-jupyter-v1-n8kzm to rexy2
  28m       28m     1   kubelet, rexy2                      Normal      SuccessfulMountVolume   MountVolume.SetUp succeeded for volume "pach-disk" 
  28m       28m     1   kubelet, rexy2                      Normal      SuccessfulMountVolume   MountVolume.SetUp succeeded for volume "docker" 
  28m       28m     1   kubelet, rexy2                      Normal      SuccessfulMountVolume   MountVolume.SetUp succeeded for volume "pach-bin" 
  28m       28m     1   kubelet, rexy2                      Normal      SuccessfulMountVolume   MountVolume.SetUp succeeded for volume "pachyderm-worker" 
  28m       28m     1   kubelet, rexy2                      Normal      SuccessfulMountVolume   MountVolume.SetUp succeeded for volume "pachyderm-storage-secret" 
  28m       28m     1   kubelet, rexy2                      Normal      SuccessfulMountVolume   MountVolume.SetUp succeeded for volume "pachyderm-token-82jpz" 
  28m       28m     1   kubelet, rexy2      spec.initContainers{init}   Normal      Pulled          Container image "pachyderm/worker:local" already present on machine
  28m       28m     1   kubelet, rexy2      spec.initContainers{init}   Normal      Created         Created container
  28m       28m     1   kubelet, rexy2      spec.initContainers{init}   Normal      Started         Started container
  28m       28m     1   kubelet, rexy2      spec.containers{user}       Normal      Pulling         pulling image "jupyter/datascience-notebook"
  14m       14m     1   kubelet, rexy2      spec.containers{user}       Normal      Pulled          Successfully pulled image "jupyter/datascience-notebook"
  14m       14m     1   kubelet, rexy2      spec.containers{storage}    Normal      Pulled          Container image "pachyderm/pachd:local" already present on machine
  14m       14m     1   kubelet, rexy2      spec.containers{storage}    Normal      Created         Created container
  14m       14m     1   kubelet, rexy2      spec.containers{storage}    Normal      Started         Started container
  14m       14m     3   kubelet, rexy2      spec.containers{user}       Normal      Created         Created container
  14m       14m     3   kubelet, rexy2      spec.containers{user}       Normal      Started         Started container
  14m       13m     4   kubelet, rexy2      spec.containers{user}       Normal      Pulled          Container image "jupyter/datascience-notebook" already present on machine
  14m       8m      29  kubelet, rexy2      spec.containers{user}       Warning     BackOff         Back-off restarting failed container
  14m       3m      51  kubelet, rexy2                      Warning     FailedSync      Error syncing pod
```

And it seems to run fine as user 1000 directly under Docker:

```
$ docker run -it jupyter/datascience-notebook bash
jovyan@ae1f3640b755:~$ whoami
jovyan
jovyan@ae1f3640b755:~$ id -u `whoami`
1000
jovyan@492e17b43f76:~$ cat /etc/passwd
root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
sys:x:3:3:sys:/dev:/usr/sbin/nologin
sync:x:4:65534:sync:/bin:/bin/sync
games:x:5:60:games:/usr/games:/usr/sbin/nologin
man:x:6:12:man:/var/cache/man:/usr/sbin/nologin
lp:x:7:7:lp:/var/spool/lpd:/usr/sbin/nologin
mail:x:8:8:mail:/var/mail:/usr/sbin/nologin
news:x:9:9:news:/var/spool/news:/usr/sbin/nologin
uucp:x:10:10:uucp:/var/spool/uucp:/usr/sbin/nologin
proxy:x:13:13:proxy:/bin:/usr/sbin/nologin
www-data:x:33:33:www-data:/var/www:/usr/sbin/nologin
backup:x:34:34:backup:/var/backups:/usr/sbin/nologin
list:x:38:38:Mailing List Manager:/var/list:/usr/sbin/nologin
irc:x:39:39:ircd:/var/run/ircd:/usr/sbin/nologin
gnats:x:41:41:Gnats Bug-Reporting System (admin):/var/lib/gnats:/usr/sbin/nologin
nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin
systemd-timesync:x:100:102:systemd Time Synchronization,,,:/run/systemd:/bin/false
systemd-network:x:101:103:systemd Network Management,,,:/run/systemd/netif:/bin/false
systemd-resolve:x:102:104:systemd Resolver,,,:/run/systemd/resolve:/bin/false
systemd-bus-proxy:x:103:105:systemd Bus Proxy,,,:/run/systemd:/bin/false
_apt:x:104:65534::/nonexistent:/bin/false
jovyan:x:1000:100::/home/jovyan:/bin/bash
messagebus:x:105:106::/var/run/dbus:/bin/false
```

**Comment by ztj** [comment]
I’ve worked around this issue and some others. My first issue is the same as the one you describe here. The output is identical. The solution is easy. Instead of directly referencing the jupyter/datascience-notebook container, make your own with a couple extra parameters. My Dockerfile looks like this:

```
FROM jupyter/datascience-notebook:9f9e5ca8fe5a

USER root
```

The next problems that I ran into have to do with port mapping and permissions. Despite specifying the correct port values for internal_port and external_port (8888, and 30888 respectively) the mapping that appeared in k8s was 30888:30888 and as a result, nothing could be accessed. Furthermore the jovyan user can not access /pfs/** by default. 

I worked around both issues by using a pipeline definition like this one...

```
{
    "pipeline": {
        "name" : "jupyterdn"
    },
    "transform": {
        "image": “local/jupyter-ds-n:1.0.0",
        "cmd": ["/bin/bash"],
        "stdin": [
            "chown -R 1000:100 /pfs",
            "start-notebook.sh --NotebookApp.password='sha1:ddddddd:dd278ddddddddddddf926af9968' --NotebookApp.port=30888"
        ]
    },
    "parallelism_spec": {
        "constant": 1
    },
    "input": {
        "atom": {
            "repo": “myinputrepo",
            "glob": "/"
        }
    },
    "service": {
        "external_port": 30888,
        "internal_port": 30888
    }
}
```

This works out okay given the parameters I used in the Dockerfile above for the container image and the fact that start-notebook.sh will downgrade the running user to jovyan during initialization and prior to running the actual server. 

So to reiterate for the sake of the issue, the problems I encountered were:

- The jupyter container appears to expect a user with uid 1000 to exist in the host context but that can be overridden and is probably not a pachyderm bug specifically
- The permissions of the /pfs directories aren’t compatible with the runtime user of the jupyter notebook server which is again not a pachyderm flaw and I showed how to deal with it one way up above
- The port mapping doesn’t work correctly and this does appear to be a pachyderm bug


Beyond the modifications I made, I’m still missing a few things to make use of this in an effective manner. For example, it would be great if there was some way we could persist notebooks/assets created by the server into a repo as well as access past notebooks in future pipeline/service launches. If there is a way and it’s not really awkward, I’d be interested to know about it. It’d be great to finish up a notebook, stop the service and have that as part of the data provenance graph as well as available to start from the next time the service launched.

**Comment by gabrielgrant** [comment]
@ztj this is super useful, thanks!

**Comment by ztj** [comment]
@gabrielgrant So to be clear, I’m preeeetty new to the container world and docker specifically, but, my understanding of how the isolation works, how the linux kernel works, and so on is such that I believe the UIDs are shared between the host running the containers and the containers themselves while the actual user name and its permissions at runtime within the container may differ from the host. At the very least, at container execution time, whatever UID is specified has to exist in the real kernel. That’s my understanding, it could be wrong, but, it jives with everything I’ve seen regarding this issue as well as things like file permissions on mounted volumes, etc.

As a result, if you try to run a container with USER 1000 specified in the Dockerfile (which is the case by default with jupyter/datascience-notebook) and the host doesn’t even have a UID 1000, it will fail. And I guess this isn’t normally an issue since most host systems generally have a user with uid 1000. Looking at it a bit more closely, it seems another solution might be to set the NB_UID and NB_GID environment variables to valid UID/GID values at runtime, but, I’m not sure if that’s something that can be done trying to use the image directly from Pachyderm, or if the actual host OS would be predictable enough or promise a particular UID would exist.

So I just took the simpler approach of running the container as UID 0 like most containers seem to be run.

As for the last question, in my one real use case so far, what I’m actually doing now is creating the jupyter pipeline when I want to explore the data in pachyderm, waiting for it to start, uploading a notebook I had downloaded from the previous run and using that. Once I’m done there, I download that notebook again and delete the pipeline until I need it again. It works okay. But it would certainly be more convenient, especially when actively working on/in a pipeline, if that series of manual actions could be skipped. Having the notebook as part of the data provenance would be an extra bonus as well, if it were possible.

That said, I suspect it’s a non-trivial matter and that other use cases for notebooks might drive a desire for different behavior entirely. And it certainly isn’t standing in my way for using pachyderm for its intended purpose.

**Comment by gabrielgrant** [comment]
@jdoliner you were working on some fixes related to this. Did those get merged in? (quick glance through the PRs didn't turn up anything obviously related, but I don't recall exactly what the root causes of the issues that you fixed were, so may have missed it)

**Comment by gabrielgrant** [comment]
To summarize there seem to be three issues:

1. User doesn't get set correctly (`user 1000 not found`)
2. non-root owners can't access /pfs
3. incorrect port mapping

1. is addressed by #3085

**Comment by ysimonson** [comment]
Closing this:

* Enterprise users can use our official jupyterhub-pachyderm service, which provides seamless integration between pachyderm and jupyterhub.
* While we don't maintain an open source version of jupyter running as a pachyderm pipeline service, we do have a demonstration of using OSS pachyderm + python-pachyderm with jupyter here: https://github.com/pachyderm/python-pachyderm/tree/master/examples/jupyter
